packages/flutter/lib/src/rendering/semantics.dart

```diff
@@ -0,0 +1,446 @@
+  void set hasCheckedState(bool value) => _flagSetter(_SemanticFlags.hasCheckedState, value);
+
+  bool get isChecked => _flags[_SemanticFlags.isChecked];
+  void set isChecked(bool value) => _flagSetter(_SemanticFlags.isChecked, value);
+
+  String get label => _label;
+  String _label = '';
+  void set label(String value) {
+    assert(value != null);
+    if (_label != value) {
+      _label = value;
+      _markDirty();
+    }
+  }
+
+  void reset() {
+    _flags.reset();
+    _label = '';
+    _markDirty();
+  }
+
+  List<SemanticsNode> _newChildren;
+  void addChildren(Iterable<SemanticsNode> children) {
+    _newChildren ??= <SemanticsNode>[];
+    _newChildren.addAll(children);
+    // we do the asserts afterwards because children is an Iterable
+    // and doing the asserts before would mean the behaviour is
+    // different in checked mode vs release mode (if you walk an
+    // iterator after having reached the end, it'll just start over;
+    // the values are not cached).
+    assert(!_newChildren.any((SemanticsNode child) => child == this));
+    assert(() {
+      SemanticsNode ancestor = this;
+      while (ancestor.parent is SemanticsNode)
+        ancestor = ancestor.parent;
+      assert(!_newChildren.any((SemanticsNode child) => child == ancestor));
+      return true;
+    });
+    assert(() {
+      Set<SemanticsNode> seenChildren = new Set<SemanticsNode>();
+      for (SemanticsNode child in _newChildren)
+        assert(seenChildren.add(child)); // check for duplicate adds
+      return true;
+    });
+  }
+
+  List<SemanticsNode> _children;
+  bool get hasChildren => _children?.isNotEmpty ?? false;
+  bool _needsCleanup = false;
+  void finalizeChildren() {
+    if (_children != null) {
+      for (SemanticsNode child in _children)
+        child._needsCleanup = true;
+    }
+    if (_newChildren != null) {
+      for (SemanticsNode child in _newChildren)
+        child._needsCleanup = false;
+    }
+    bool sawChange = false;
+    if (_children != null) {
+      for (SemanticsNode child in _children) {
+        if (child._needsCleanup) {
+          if (child.parent == this) {
+            // we might have already had our child stolen from us by
+            // another node that is deeper in the tree.
+            dropChild(child);
+          }
+          sawChange = true;
+        }
+      }
+    }
+    if (_newChildren != null) {
+      for (SemanticsNode child in _newChildren) {
+        if (child.parent != this) {
+          if (child.parent != null) {
+            // we're rebuilding the tree from the bottom up, so it's possible
+            // that our child was, in the last pass, a child of one of our      
+            // ancestors. In that case, we drop the child eagerly here.
+            // TODO(ianh): Find a way to assert that the same node didn't
+            // actually appear in the tree in two places.
+            child.parent?.dropChild(child);
```

Comment:
I really don't like that we're building the tree upside down. It'd be much cleaner if it built it straight down. I think this might be part of the cause of the bug where things are getting lost (drop the popup menu, cancel it, drop it again, the box that would be getting the cancel isn't included the second time).